proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+  type KnownRGB :: (Nat, Nat, Nat) -> Constraint
+  class KnownRGB c where
+    _rgbVal :: (Word8, Word8, Word8)
+
+  rgb :: forall s -> KnownRGB (ParseRGB c) => (Word8, Word8, Word8)
+  rgb s = _rgbVal @(ParseRGB s)
+
+Use site::
+
+  ghci> rgb "red"
+  (255, 0, 0)
+
+  ghci> rgb "#112233"
+  (17, 34, 51)
+
+  ghci> rgb "asdfasdf"
+  -- custom type error from ParseRGB
+
+Effect and Interactions
+-----------------------
+
+* Visible ``forall`` becomes available in types of terms, making them more similar to
+  types of types. There remains a discrepancy that ``forall`` for types is actually a
+  relevant quantifier, while the proposed ``forall x ->`` for terms is
+  irrelevant. This is to be resolved in the future by making type-level
+  ``forall`` irrelevant.
+
+* The renaming of a visible dependent argument is different than that of a
+  dependent argument with a visibility override. Consider this code::
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    data T = T
+
+    a = f @T
+    b = g  T
+
+  In ``f @T``, we refer to the type constructor, but in ``g T`` we refer to the
+  data constructor.
+
+  The implementation may offer warning flags to help the user identify such
+  ambiguous occurrences.
+
+* When punned names come from external libraries, there are several workarounds
+  to resolve the ambiguity:
+
+  1. Using ``-XExplicitNamespaces``::
+
+      import Data.Proxy
+      import qualified Data.Proxy (type Proxy) as T
+
+      x = f   Proxy  -- refers to the data constructor
+      y = f T.Proxy  -- refers to the type constructor
+
+  2. Using a type synonym::
+
+      import Data.Proxy
+
+      type TProxy = Proxy
+
+      x = f  Proxy  -- refers to the data constructor
+      y = f TProxy  -- refers to the type constructor
+
+  We may offer a more convenient way to disambiguate namespaces in the future,
+  based on the experience of using the proposed feature.
+
+* Identifiers bound to terms are
+  not promoted. Consider this well-typed program::
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    a = f @(Just True)  -- ok
+    b = g  (Just True)  -- ok
+
+  If we factor out ``Just True`` into a type synonym, it continues to work::
+
+    type X :: Maybe Bool
+    type X = Just True
+
+    a = f @X  -- still OK
+    b = g  X  -- still OK
+
+  However, if we bind it to a term-level variable, the example becomes
+  ill-typed::
+
+    x :: Maybe Bool
+    x = Just True
+
+    a = f @x  -- not currently valid
+    b = g  x  -- not valid under the proposal
+
+  This is because we retain the distinction between terms and types. This
+  proposal is a step towards dependent types, but it does not go all the way.
+  Accepting the program above is left as future work.
+
+Costs and Drawbacks
+-------------------
+
+This is one more feature to implement and maintain.
+
+Alternatives
+------------
+
+1.  Keep types and terms forever different by not supporting visible ``forall``
+    in terms.
```

Comment:
You can remove this line: it's self-evident that proposal rejection is an option.

Reply:
It's there to bring the reader's attention to the consequences of rejecting this proposal, i.e. that types and terms will not be unified unless we figure out how to allow `forall x -> t` in types of terms. Maybe it's obvious to the committee members, but not to readers with other background.